Documentation/proposals/202405-agent-daemonset.md

```diff
@@ -0,0 +1,133 @@
+# DaemonSet deployment for Prometheus Agent
+
+* Owners:
+  * [haanhvu](https://github.com/haanhvu)
+* Related Tickets:
+  * [#5495](https://github.com/prometheus-operator/prometheus-operator/issues/5495)
+* Other docs:
+  * n/a
+
+This proposal is about designing and implementing the deployment of Prometheus Agent as DaemonSet. Currently, Prometheus Agent can only be deployed as StatefulSet, which could be considered as “cluster-wide” strategy, meaning one or several high-availability Prometheus Agents are responsible for scraping metrics of the whole cluster. Though this works well for many use cases, some use cases may indeed prefer a “node-specific” strategy (DaemonSet), where each Prometheus Agent only scrapes the metrics from the targets located on the same node.
+
+
+## 1. Why
+
+When deploying Prometheus Agent in Kubernetes, three of the biggest users’ concerns are: load distribution, scalability, and security.
```

Comment:
I didn't cite it from a source. I read users' blogs, browsed the issues in our repo, and from my experience of setting up some nodes for benchmark pipelines for Jaeger in last year GSoC to form this general observation.

There're of course other concerns though, like cost^^ But I don't think DaemonSet helps anything with reducing cost, so didn't state it here.

Reply:
> In some way it helps a lot with the cost, because it scales with the load, so you don't need to keep buffy Prometheus servers when your cluster is scaled back

Yeah I mentioned automatic scaleup, but forgot to mention automatic scaledown. I'll add this to Scalability section then. We don't have any proof of cost so I wouldn't mention cost here. But automatic scaledown implicitly refer to cost optimization.